src/client.js

```diff
@@ -0,0 +1,229 @@
+        txtRegex: /^"?cabalkey=([0-9a-f]{64})"?$/i,
+    }
+    // also takes opts.persistentCache which has a read and write function
+    //   read: async function ()   // aka cache lookup function
+    //   write: async function ()  // aka cache write function
+    if (opts.persistentCache) cabalDnsOpts.persistentCache = opts.persistentCache
+    this.cabalDns = DatDns(cabalDnsOpts)
+  }
+
+  static getDatabaseVersion () {
+    return Cabal.databaseVersion
+  }
+
+  static scrub (key) {
+    return key.replace('cabal://', '').replace('cbl://', '').replace('dat://', '').replace(/\//g, '')
+  }
+
+  resolveName (name, cb) {
+      return this.cabalDns.resolveName(name).then((key) => { 
+          if (!cb) return Client.scrub(key)
+          cb(Client.scrub(key)) 
+      })
+  }
+
+  createCabal () {
+    return this.addCabal(crypto.keyPair().publicKey.toString('hex'))
+  }
+
+  addCabal (key) {
+      let cabalPromise
+      if (typeof key === 'string') {
+          cabalPromise = this.resolveName(key).then((resolvedKey) => {
+            const {temp, dbdir} = this.config
+            const storage = temp ? ram : dbdir + resolvedKey
+            if (!temp) try { mkdirp.sync(path.join(dbdir, resolvedKey, 'views')) } catch (e) {}
+            var db = temp ? memdb() : level(path.join(dbdir, resolvedKey, 'views'))
+            var cabal = Cabal(storage, resolvedKey, {db: db, maxFeeds: this.maxFeeds})
+            this._keyToCabal[resolvedKey] = cabal
+            return cabal
+          })
+      } else {
+          cabalPromise = new Promise((res, rej) => {
+            // a cabal instance was passed in
+            var cabal = key
+            this._keyToCabal[cabal.key] = cabal
+            res(cabal)
+          })
+      }
+      return new Promise((res, rej) => {
+          cabalPromise.then((cabal) => {
+              if (!this.currentCabal) {
+                  this.currentCabal = cabal
+              }
+              cabal.ready(() => {
+                  const details = new CabalDetails(cabal)
+                  this.cabals.set(cabal, details)
+                  cabal.swarm()
+                  this.getCurrentCabal()._emitUpdate()
+                  res(details)
+              })
+          })
+    })
+  }
+
+  focusCabal (key) {
+    const cabal = this._coerceToCabal(key)
+    if (!cabal) {
+      return false
+    }
+    this.currentCabal = cabal
+    return this.cabalToDetails(cabal)
+  }
+
+  removeCabal (key) {
+    const cabal = this._coerceToCabal(key)
+    if (!cabal) {
+      return false
+    }
+
+    const details = this.cabalToDetails(cabal)
+    details._destroy()
+
+    // burn everything we know about the cabal
+    this._keyToCabal[key] = null
+    return this.cabals.delete(cabal)
+  }
+
+  getCabalKeys () {
+    return Object.keys(this._keyToCabal) // ???: sorted?
+  }
+
+  getCurrentCabal () {
+    return this.cabalToDetails(this.currentCabal)
+  }
+
+  getCabalByKey (key) {
```

Comment:
It feels a little confusing to me to have `getCurrentCabal` return a CabalDetails, but `getCabalByKey` return a `Cabal`. Do we need to expose the underlying `Cabal`(Core) instance to the API? This would also remove the need to expose `cabalToDetails` to the outside world.

Reply:
ideally not, but i don't know if someone wants to do some weird mumbojumbo in their client that we haven't added support for yet in the `cabal-client`, and this method would support that use case with very little other scope creep.

i can remove it if you still think it's superfluous; i agree it's a bit confusing name-wise